ARROW-10402: [Rust] Refactor array equality

jorgecarleitao opened the pull request:
This is a major refactor of the `equal.rs` module.

The rational for this change is many fold: 

* currently array comparison requires downcasting the array ref to its concrete types. This is painful and not very ergonomics, as the user must "guess" what to downcast for comparison. We can see this in the hacks around `sort`, `take` and `concatenate` kernel's tests, and some of the tests of the builders.
* the code in array comparison is difficult to follow given the amount of calls that they perform around offsets.
* The implementation currently indirectly uses many of the `unsafe` APIs that we have (via pointer aritmetics), which makes it risky to operate and mutate.
* Some code is being repeated.

This PR:

1. adds `impl PartialEq for dyn Array`, to allow `Array` comparison based on `Array::data` (main change)
2. Makes array equality to only depend on `ArrayData`, i.e. it no longer depends on concrete array types (such as `PrimitiveArray` and related API) to perform comparisons.
3. Significantly reduces the risk of panics and UB when composite arrays are of different types, by checking the types on `range` comparison
4. Makes array equality be statically dispatched, via `match datatype`.
5. DRY the code around array equality
6. Fixes an error in equality of dictionary with equal values
7. Added tests to equalities that were not tested (fixed binary, some edge cases of dictionaries)
8. splits `equal.rs` in smaller, more manageable files.
9. Removes `ArrayListOps`, since it it no longer needed
10. Moves Json equality to its own module, for clarity.
11. removes the need to have two functions per type to compare arrays.
12. Adds the number of buffers and their respective width to datatypes from the specification. This was backported from #8401
13. adds a benchmark for array equality

Note that this does not implement `PartialEq` for `ArrayData`, only `dyn Array`, as different data does not imply a different array (due to nullability). That implementation is being worked on #8200.

IMO this PR significantly simplifies the code around array comparison, to the point where many implementations are 5 lines long.

This also improves performance by 10-40%.

<details>
 <summary>Benchmark results</summary>

```
Previous HEAD position was 3dd3c69d7 Added bench for equality.
Switched to branch 'equal'
Your branch is up to date with 'origin/equal'.
   Compiling arrow v3.0.0-SNAPSHOT (/Users/jorgecarleitao/projects/arrow/rust/arrow)
    Finished bench [optimized] target(s) in 51.28s
     Running /Users/jorgecarleitao/projects/arrow/rust/target/release/deps/equal-176c3cb11360bd12
Gnuplot not found, using plotters backend
equal_512               time:   [36.861 ns 36.894 ns 36.934 ns]                       
                        change: [-43.752% -43.400% -43.005%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  7 (7.00%) high mild
  5 (5.00%) high severe

equal_nulls_512         time:   [2.3271 us 2.3299 us 2.3331 us]                             
                        change: [-10.846% -9.0877% -7.7336%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe

equal_string_512        time:   [49.219 ns 49.347 ns 49.517 ns]                              
                        change: [-30.789% -30.538% -30.235%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe

equal_string_nulls_512  time:   [3.7873 us 3.7939 us 3.8013 us]                                    
                        change: [-8.2944% -7.0636% -5.4266%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe
```

</details>

All tests are there, plus new tests for some of the edge cases and untested arrays.

This change is backward incompatible `array1.equals(&array2)` no longer works: use `array1 == array2` instead, which is the idiomatic way of comparing structs and trait objects in rust.

nevi-me commented on rust/arrow/src/array/data.rs:
```diff
@@ -207,6 +210,56 @@ impl ArrayData {
+    }
 }
 
 impl PartialEq for ArrayData {
```
My changes here muddy the water. While reviewing this, I started to clean this up. I'll try submit a PR on top of this, removing the asserts, and completing the implementation.

jorgecarleitao replied:
One Idea would be to use the equality code from this PR, but sprinkle it with `debug_assert_eq`, so that during debug, it is more clear where the difference is. An alternative is to use `assert_eq`, so that the two arrays are shown side-by-side (this only works for small arrays, though).

nevi-me commented on rust/arrow/src/array/equal/variable_size.rs:
```diff
@@ -0,0 +1,90 @@
+        )
+}
+
+pub fn variable_sized_equal<T: OffsetSizeTrait>(
```
May you please add a doc, as it's a public function

jorgecarleitao replied:
I have privatized this function, as it assumes that the user already used `base_equal`. This reminded me to document the `equal::equal` function, which is the public one of the module.

jorgecarleitao commented on rust/arrow/src/array/equal/null.rs:
```diff
@@ -0,0 +1,31 @@
+) -> bool {
+    // a null buffer's range is always true, as every element is by definition equal (to null).
+    // We only need to compare data_types
+    true
```
Note that this function already assumes that me performed that comparison before-hand. We do that on the `equal::equal` and `equal::equal_range` functions.

The reason I moved `base_equal` away from type-specific functions is that it allow us to have a single function per type for both range equality and equality.

To avoid the risk of being misused (i.e. not checking `base_equal` first), I pushed a commit that privatizes all these functions, effectively making `equal::equal` the only public function of the module `equal`.

jorgecarleitao replied:
I added a test for this, just for avoidance of doubts :)

jorgecarleitao commented:
I have simplified this code further to allow the compiler to optimize out some functions. The code is now 10-40% faster compared to current master. I updated the description accordingly.

nevi-me commented on rust/arrow/src/array/equal/structure.rs:
```diff
@@ -0,0 +1,59 @@
+    lhs_start: usize,
+    rhs_start: usize,
+    len: usize,
+) -> bool {
```
Something that we can address later, is that if a struct slot is null, we should carry that nullness across to its children. This is actually the problem that @carols10cents and I encountered on both parquet roundtrips and integration tests.

The spec (http://arrow.apache.org/docs/format/Columnar.html#struct-layout) says:

> While a struct does not have physical storage for each of its semantic slots (i.e. each scalar C-like struct), an entire struct slot can be set to null via the validity bitmap. Any of the child field arrays can have null values according to their respective independent validity bitmaps. This implies that for a particular struct slot the validity bitmap for the struct array might indicate a null slot when one or more of its child arrays has a non-null value in their corresponding slot. When reading the struct array the parent validity bitmap takes priority.

I think we can address this in a separate PR, if we still find that `struct_equal` doesn't cover that edge case.

nevi-me reviewed (approved):
Thanks @jorgecarleitao 

I'm happy that this PR's been updated for logical equality, and I'd propose we merge it in, then look at #8590 and #8200 to address edge-cases where the IPC and Parquet tests still fail equality tests.

@carols10cents @alamb @paddyhoran @jhorstmann  what's your opinion?

alamb commented on rust/arrow/src/array/data.rs:
```diff
@@ -207,6 +206,41 @@ impl ArrayData {
+        new_data
+    }
+
+    /// Returns the buffer `buffer` as a slice of type `T`. The slice is already offset.
```
```suggestion
    /// Returns self.buffers at index `buffer` as a slice of type `T` starting at self.offset
```

alamb commented on rust/arrow/src/array/equal/mod.rs:
```diff
@@ -0,0 +1,829 @@
+}
+
+#[inline]
+fn equal_values(
```
```suggestion
/// Compares two arrays for equality starting at `lhs_start` and `rhs_start`
/// `lhs` and `rhs` *must* have the same data type
fn equal_values(
```

alamb reviewed (approved):
I can't say I made it through this entire PR carefully, but I reviewed about 50% of it carefully and skimmed the rest -- what I did see had a logical and understandable structure I would feel comfortable supporting / maintaining, FWIW. 

I agree with @jorgecarleitao  that having a uniform definition of array equality will solve many of the challenges we have seen (especially with parquet <--> arrow round tripping)

Thus my opinion is that we should merge this.

jhorstmann commented on rust/arrow/src/array/data.rs:
```diff
@@ -121,7 +120,7 @@ impl ArrayData {
     pub fn is_null(&self, i: usize) -> bool {
         if let Some(ref b) = self.null_bitmap {
-            return !b.is_set(i);
+            return !b.is_set(self.offset + i);
```
Thanks a lot, I think I have another ticket that would be fixed by this change

jorgecarleitao commented:
@alamb , @nevi-me and @jhorstmann , thanks a lot for taking a close look into this, also for the amazing work yesterday and today on reviewing and merging stuff. Really impressive! 💯 

I rebased this and resolved conflicts.

nevi-me commented:
> @alamb , @nevi-me and @jhorstmann , thanks a lot for taking a close look into this, also for the amazing work yesterday and today on reviewing and merging stuff. Really impressive! 💯
> 
> I rebased this and resolved conflicts.

I'm super-excited, because this likely unlocks us on the failures that I've been getting on #8200.
